drivers/freebsd/tcpsrc/tcpsrc.4.txt
NAME
  tcpsrc - TCP client socket source
SYNOPSIS
  tcpsrc
DESCRIPTION
  the tcpsrc device allows userspace applications to establish
  TCP connections using calls to ioctl(2) on a file descriptor previously
  created by a call to open(2) on the device.
RATIONALE
  Using standard POSIX APIs to establish non-blocking TCP client sockets
  require three syscalls: socket(2), fcntl(2) and connect(2). Setting
  both O_NONBLOCK and O_CLOEXEC requires two calls to fcntl(2) per socket.

  Using two widely available options to socket(2): SOCK_NONBLOCK and
  SOCK_CLOEXEC brings the number of syscalls down to two: socket(2) and
  connect(2).

  In a sandboxed environment, it is often times impossible or unwise to
  have direct access to these syscalls.

  One way to make it possible for untrusted processes to establish TCP
  connections while sandboxed is to open up IPC with a trusted process before
  entering sandbox mode. After having entered sandbox mode, the untrusted
  process may then request connections from the trusted process, getting the
  file descriptors back using file descriptor passing. The trusted process
  may then choose which addresses the untrusted process are allowed to connect
  to, or apply any other policies; e.g., the rate of establishing connections
  or the number of allowed connections.

  One downside with such an approach is the added overhead of the IPC.

  By providing a standard UNIX pseudo-device, tcpsrc attempts to both limit the
  number of syscalls needed to establish multiple new TCP connections, as well
  as providing a suitable interface for sandboxed processes. A sandboxed
  process can obtain a file descriptor to the tcpsrc device and use it to
  create new TCP connections. This file descriptor can have a set of
  limitations associated with it before entering sandbox mode or before the
  file descriptor is passed to an untrusted process.
IOCTLS
  All tcpsrc ioctls are defined in:

	#include <yans/tcpsrc.h>

    TCPSRC_CONNECT  (struct tcpsrc_conn)  Establish a TCP connection.
    TCPSRC_ALLOW    (struct tcpsrc_range) Explicitly whitelist a range.
    TCPSRC_DISALLOW (struct tcpsrc_range) Explicitly blacklist a range.

  All tcpsrc ioctls return -1 on error and sets errno accordingly.

  On success, TCPSRC_CONNECT returns a valid file descriptor. TCPSRC_ALLOW
  and TCLSRC_DISALLOW returns 0.


EXAMPLE
  This example opens the tcpsrc device and opens a TCP connection to localhost.

	#include <sys/socket.h>
	#include <arpa/inet.h>
	#include <unistd.h>
	#include <stdlib.h>
	#include <stdio.h>
	#include <sys/capsicum.h>
	
	#include <yans/tcpsrc.h>
	
	int main() {
	  int fd;
	  int clifd;
	  int result = EXIT_FAILURE;
	  struct tcpsrc_conn conn = {{{0}}};
	
	  fd = open("/dev/tcpsrc", O_RDWR);
	  if (fd < 0) {
	    perror("open");
	    goto out;
	  }
	
	  cap_enter();
	
	  conn.u.sin.sin_family = AF_INET;
	  conn.u.sin.sin_addr.s_addr = inet_addr("127.0.0.1");
	  conn.u.sin.sin_port = htons(22);
	  clifd = ioctl(fd, TCPSRC_CONNECT, &conn);
	  if (clifd < 0) {
	    perror("TCPSRC_CONNECT");
	    goto out_close;
	  }
	
	  close(clifd);
	  result = EXIT_SUCCESS;
	out_close:
	  close(fd);
	out:
	  return result;
	}

COMPARISON
  1. socket(2)/connect(2)

	kqueue()			 = 3 (0x3)
	socket(PF_INET,SOCK_STREAM|SOCK_NONBLOCK, IPPROTO_TCP) = 4 (0x4)
	connect(4,{ ... },16) ERR#36 'Operation now in progress'
	socket(PF_INET,SOCK_STREAM|SOCK_NONBLOCK,IPPROTO_TCP) = 5 (0x5)
	connect(5,{ ... },16) ERR#36 'Operation now in progress'
	socket(PF_INET,SOCK_STREAM|SOCK_NONBLOCK,IPPROTO_TCP) = 6 (0x6)
	connect(6,{ ... },16) ERR#36 'Operation now in progress'
	socket(PF_INET,SOCK_STREAM|SOCK_NONBLOCK,IPPROTO_TCP) = 7 (0x7)
	connect(7,{ ... },16) ERR#36 'Operation now in progress'
	socket(PF_INET,SOCK_STREAM|SOCK_NONBLOCK,IPPROTO_TCP) = 8 (0x8)
	connect(8,{ ... },16) ERR#36 'Operation now in progress'
	socket(PF_INET,SOCK_STREAM|SOCK_NONBLOCK,IPPROTO_TCP) = 9 (0x9)
	connect(9,{ ... },16) ERR#36 'Operation now in progress'
	socket(PF_INET,SOCK_STREAM|SOCK_NONBLOCK,IPPROTO_TCP) = 10 (0xa)
	connect(10,{ ... },16) ERR#36 'Operation now in progress'
	socket(PF_INET,SOCK_STREAM|SOCK_NONBLOCK,IPPROTO_TCP) = 11 (0xb)
	connect(11,{ ... },16) ERR#36 'Operation now in progress'
	socket(PF_INET,SOCK_STREAM|SOCK_NONBLOCK,IPPROTO_TCP) = 12 (0xc)
	connect(12,{ ... },16) ERR#36 'Operation now in progress'
	socket(PF_INET,SOCK_STREAM|SOCK_NONBLOCK,IPPROTO_TCP) = 13 (0xd)
	connect(13,{ ... },16) ERR#36 'Operation now in progress'
	socket(PF_INET,SOCK_STREAM|SOCK_NONBLOCK,IPPROTO_TCP) = 14 (0xe)
	connect(14,{ ... },16) ERR#36 'Operation now in progress'
	kevent(3, {...}, 22, {...}, 64, { 0.050000000 }) = 1 (0x1)

    tcpdump with timestamps being delta between packets (-ttt):

	00:00:00.000000 IP ... > ...: Flags [S], ...
	00:00:00.000048 IP ... > ...: Flags [S], ...
	00:00:00.000028 IP ... > ...: Flags [S], ...
	00:00:00.000025 IP ... > ...: Flags [S], ...
	00:00:00.000025 IP ... > ...: Flags [S], ...
	00:00:00.000025 IP ... > ...: Flags [S], ...
	00:00:00.000025 IP ... > ...: Flags [S], ...
	00:00:00.000025 IP ... > ...: Flags [S], ...
	00:00:00.000025 IP ... > ...: Flags [S], ...
	00:00:00.000025 IP ... > ...: Flags [S], ...
	00:00:00.000025 IP ... > ...: Flags [S], ...

  2. IPC with external service (netstring serialization) and fd passing

	socket(PF_LOCAL,SOCK_STREAM,0)		 = 3 (0x3)
	connect(3,{ AF_UNIX "connector.sock" },106) = 0 (0x0)
	cap_enter()				 = 0 (0x0)
	kqueue()					 = 4 (0x4)
	write(3,"42:5:proto,1:6,7:dstaddr,16:\^P"...,46) = 46 (0x2e)
	recvmsg(3, ... ) = 4 (0x4)
	write(3,"42:5:proto,1:6,7:dstaddr,16:\^P"...,46) = 46 (0x2e)
	recvmsg(3, ... ) = 4 (0x4)
	write(3,"42:5:proto,1:6,7:dstaddr,16:\^P"...,46) = 46 (0x2e)
	recvmsg(3, ... ) = 4 (0x4)
	write(3,"42:5:proto,1:6,7:dstaddr,16:\^P"...,46) = 46 (0x2e)
	recvmsg(3, ... ) = 4 (0x4)
	write(3,"42:5:proto,1:6,7:dstaddr,16:\^P"...,46) = 46 (0x2e)
	recvmsg(3, ... ) = 4 (0x4)
	write(3,"42:5:proto,1:6,7:dstaddr,16:\^P"...,46) = 46 (0x2e)
	recvmsg(3, ... ) = 4 (0x4)
	write(3,"42:5:proto,1:6,7:dstaddr,16:\^P"...,46) = 46 (0x2e)
	recvmsg(3, ... ) = 4 (0x4)
	write(3,"42:5:proto,1:6,7:dstaddr,16:\^P"...,46) = 46 (0x2e)
	recvmsg(3, ... ) = 4 (0x4)
	write(3,"42:5:proto,1:6,7:dstaddr,16:\^P"...,46) = 46 (0x2e)
	recvmsg(3, ... ) = 4 (0x4)
	write(3,"42:5:proto,1:6,7:dstaddr,16:\^P"...,46) = 46 (0x2e)
	recvmsg(3, ... ) = 4 (0x4)
	write(3,"42:5:proto,1:6,7:dstaddr,16:\^P"...,46) = 46 (0x2e)
	recvmsg(3, ... ) = 4 (0x4)
	kevent(4, {...}, 22, {...}, 64, { 0.050000000 }) = 1 (0x1)

    tcpdump with timestamps being delta between packets (-ttt):

	00:00:00.000000 IP ... > ...: Flags [S], ...
	00:00:00.000177 IP ... > ...: Flags [S], ...
	00:00:00.000179 IP ... > ...: Flags [S], ...
	00:00:00.000151 IP ... > ...: Flags [S], ...
	00:00:00.000108 IP ... > ...: Flags [S], ...
	00:00:00.000146 IP ... > ...: Flags [S], ...
	00:00:00.000126 IP ... > ...: Flags [S], ...
	00:00:00.000111 IP ... > ...: Flags [S], ...
	00:00:00.000105 IP ... > ...: Flags [S], ...
	00:00:00.000096 IP ... > ...: Flags [S], ...
	00:00:00.000091 IP ... > ...: Flags [S], ...



  3. tcpsrc(4)

	openat(AT_FDCWD,"/dev/tcpsrc",O_RDWR,00)	 = 3 (0x3)
	cap_enter()				 = 0 (0x0)
	kqueue()					 = 4 (0x4)
	ioctl(3, { IOW 0x38('8'), 0, 28 },0x7fffffffd430) = 5 (0x5)
	ioctl(3, { IOW 0x38('8'), 0, 28 },0x7fffffffd430) = 6 (0x6)
	ioctl(3, { IOW 0x38('8'), 0, 28 },0x7fffffffd430) = 7 (0x7)
	ioctl(3, { IOW 0x38('8'), 0, 28 },0x7fffffffd430) = 8 (0x8)
	ioctl(3, { IOW 0x38('8'), 0, 28 },0x7fffffffd430) = 9 (0x9)
	ioctl(3, { IOW 0x38('8'), 0, 28 },0x7fffffffd430) = 10 (0xa)
	ioctl(3, { IOW 0x38('8'), 0, 28 },0x7fffffffd430) = 11 (0xb)
	ioctl(3, { IOW 0x38('8'), 0, 28 },0x7fffffffd430) = 12 (0xc)
	ioctl(3, { IOW 0x38('8'), 0, 28 },0x7fffffffd430) = 13 (0xd)
	ioctl(3, { IOW 0x38('8'), 0, 28 },0x7fffffffd430) = 14 (0xe)
	ioctl(3, { IOW 0x38('8'), 0, 28 },0x7fffffffd430) = 15 (0xf)
	kevent(4, {...}, 22, {...}, 64, { 0.050000000 }) = 1 (0x1)

    tcpdump with timestamps being delta between packets (-ttt):

	00:00:00.000000 IP ... > ...: Flags [S], ...
	00:00:00.000043 IP ... > ...: Flags [S], ...
	00:00:00.000024 IP ... > ...: Flags [S], ...
	00:00:00.000024 IP ... > ...: Flags [S], ...
	00:00:00.000023 IP ... > ...: Flags [S], ...
	00:00:00.000023 IP ... > ...: Flags [S], ...
	00:00:00.000029 IP ... > ...: Flags [S], ...
	00:00:00.000024 IP ... > ...: Flags [S], ...
	00:00:00.000023 IP ... > ...: Flags [S], ...
	00:00:00.000025 IP ... > ...: Flags [S], ...
	00:00:00.000025 IP ... > ...: Flags [S], ...
NOTES
  As tcpsrc is trusted code running inside the kernel, it is important for
  tcpsrc to remain small and simple.
SEE ALSO
  socket(2), fcntl(2), connect(2), open(2), ioctl(2).
AUTHOR
  Sebastian Cato <seb.cato@gmail.com>

txt2man -r yans-1.1.0 -t tcpsrc -s 4 -v "Yans user's manual" tcpsrc.4.txt > tcpsrc.4


drivers/freebsd/tcpsrc/tcpsrc.c
#include <sys/types.h>
#include <sys/module.h>
#include <sys/conf.h>
#include <sys/kernel.h>
#include <sys/socket.h>
#include <sys/socketvar.h>
#include <sys/sockopt.h>
#include <sys/proc.h>
#include <sys/fcntl.h>
#include <sys/file.h>
#include <sys/filio.h>
#include <sys/filedesc.h>
#include <netinet/tcp.h>

#include "tcpsrc.h"

#define MODULE_NAME "tcpsrc"

static d_open_t      tcpsrc_open;
static d_close_t     tcpsrc_close;
static d_ioctl_t     tcpsrc_ioctl;

static struct cdevsw tcpsrc_cdevsw = {
	.d_version = D_VERSION,
	.d_open    = tcpsrc_open,
	.d_close   = tcpsrc_close,
	.d_ioctl   = tcpsrc_ioctl,
	.d_name    = MODULE_NAME,
};

static struct cdev *tcpsrc_dev;

static int
_setsockopt(struct socket *so, int level, int name, int val)
{
	struct sockopt opt = {
		.sopt_dir = SOPT_SET,
		.sopt_level = level,
		.sopt_name = name,
		.sopt_val = &val,
		.sopt_valsize = sizeof(val),
	};

	return sosetopt(so, &opt);
}

/* create a non-blocking socket w/ O_CLOEXEC set and connect(2) it */
static int
_connect(struct thread *td, struct sockaddr *sa, int *outfd)
{
	struct socket *so;
	struct file *fp;
	int fd;
	int fflag = FNONBLOCK; /* NB: non-blocking socket by default */
	int ret;

	/* the falloc,finit,&c are defined in sys/filedesc.h and
	 * implemented in kern/kern_descrip.c. The so* functions are
	 * defined in sys/socketvar.h and implemented in
	 * kern/uipc_socket.c. Their functionality is documented there. */

	/* allocate a struct file and a file descriptor w/ O_CLOEXEC
	 * set */
	ret = falloc(td, &fp, &fd, O_CLOEXEC);
	if (ret != 0)
		return (ret);

	/* alloc and initialize a struct socket. Put it in 'so' */
	ret = socreate(sa->sa_family, &so, SOCK_STREAM, IPPROTO_TCP,
            td->td_ucred, td);
	if (ret != 0) {
		fdclose(td, fp, fd);
		goto done;
	}

	/* associate the socket with the file, set FNONBLOCK */
	finit(fp, FREAD | FWRITE | FNONBLOCK, DTYPE_SOCKET, so, &socketops);
	fo_ioctl(fp, FIONBIO, &fflag, td->td_ucred, td);

	/* Set SO_REUSEADDR to 1 - allowing us to rebind on addr:port
         * pairs in TIME_WAIT state */
	ret = _setsockopt(so, SOL_SOCKET, SO_REUSEADDR, 1);
	if (ret != 0) {
		fdclose(td, fp, fd);
		goto done;
	}

	/* Set TCP_NODELAY to 1.
         * If the connection has non-ACKed data in transit, the TCP
         * stack may buffer data passed from userspace before sending it,
	 * if the size of the data in the send buffer is less than the MSS.
	 * By turning TCP_NODELAY on, we disable this functionality
	 * (Nagle's algorithm) and data passed to the kernel will be sent
	 * ASAP.
	 * We assume that userspace does not do a lot of calls to write(2)
	 * with partial data. Userspace should put the data in a buffer
         * first and/or use writev(2) on all available, relevant data */
	ret = _setsockopt(so, IPPROTO_TCP, TCP_NODELAY, 1);
	if (ret != 0) {
		fdclose(td, fp, fd);
		goto done;
	}

	/* connect! */
	ret = soconnect(so, sa, td);
	if (ret != 0) {
		so->so_state &= ~SS_ISCONNECTING;
		fdclose(td, fp, fd);
		goto done;
	}

	*outfd = fd;
done:
	fdrop(fp, td);
	return (ret);
}

static int
tcpsrc_connect(struct thread *td, struct tcpsrc_conn *conn, int *fd)
{
	/* patch up sa_len since it's not really standard */
	if (conn->u.sa.sa_family == AF_INET) {
		conn->u.sa.sa_len = sizeof(struct sockaddr_in);
	} else if (conn->u.sa.sa_family == AF_INET6) {
		conn->u.sa.sa_len = sizeof(struct sockaddr_in6);
	}

	return _connect(td, &conn->u.sa, fd);
}

static int
tcpsrc_ioctl(struct cdev *dev, u_long cmd, caddr_t addr, int flag,
		struct thread *td)
{
	int fd;
	int ret;

	switch (cmd) {
	case TCPSRC_CONNECT:
		ret = tcpsrc_connect(td, (struct tcpsrc_conn*)addr, &fd);
		if (ret == 0)
			td->td_retval[0] = fd;
		return ret;
	}

	return (EINVAL);
}

static int
tcpsrc_open(struct cdev *dev, int oflags, int devtype, struct thread *td)
{
	return (0);
}

static int
tcpsrc_close(struct cdev *dev, int fflag, int devtype, struct thread *td)
{

	return (0);
}

static int
tcpsrc_loader(struct module *m, int what, void *arg)
{
	int error = 0;

	switch (what) {
	case MOD_LOAD:
		error = make_dev_p(MAKEDEV_CHECKNAME | MAKEDEV_WAITOK,
				&tcpsrc_dev, &tcpsrc_cdevsw, 0,
				UID_ROOT, GID_WHEEL, 0660, MODULE_NAME);
		if (error != 0)
			break;
		printf("tcpsrc: module loaded\n");
		break;
	case MOD_UNLOAD:
		destroy_dev(tcpsrc_dev);
		printf("tcpsrc: module unloaded\n");
		break;
	default:
		error = EOPNOTSUPP;
		break;
	}
	return (error);
}

DEV_MODULE(tcpsrc, tcpsrc_loader, NULL);
